sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestUseTemplateEngine.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.sqlobject;
+
+import org.jdbi.v3.core.Jdbi;
+import org.jdbi.v3.core.rule.DatabaseRule;
+import org.jdbi.v3.core.rule.SqliteDatabaseRule;
+import org.jdbi.v3.core.statement.DefinedAttributeTemplateEngine;
+import org.jdbi.v3.core.statement.MessageFormatTemplateEngine;
+import org.jdbi.v3.sqlobject.config.UseTemplateEngine;
+import org.jdbi.v3.sqlobject.customizer.Define;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestUseTemplateEngine {
+    @Rule
+    public DatabaseRule db = new SqliteDatabaseRule().withPlugin(new SqlObjectPlugin());
+
+    private Jdbi jdbi;
+
+    @Before
+    public void before() {
+        jdbi = db.getJdbi();
+    }
+
+    @Test
+    public void testUseMessageFormat() {
+        String selected = jdbi.withExtension(QueriesForMessageFormatTE.class, q -> q.select("foo"));
+
+        assertThat(selected).isEqualTo("foo");
+    }
+
+    @Test
+    public void testUseDefinedAttributes() {
+        String selected = jdbi.withExtension(QueriesForDefinedAttributeTE.class, q -> q.select("foo"));
+
+        assertThat(selected).isEqualTo("foo");
+    }
+
+    public interface QueriesForMessageFormatTE {
+        @UseTemplateEngine(MessageFormatTemplateEngine.class)
```

Comment:
I can't see a valid reason why `MessageFormatTemplateEngine` should not have a public no-arg constructor. It has no state and nothing about it is configurable.

Reply:
Maybe it's a good idea to establish a project convention, constructors vs consts and factory methods? I can live with either but I agree having a mix of both is kind of annoying.

Reply:
Maybe an answer to a different question will help me decide on how I design my classes for jdbi.

Suppose I write a class, give it a public default constructor for convenience like with most classes in jdbi. In the future, something is changed about the class and having a public no-arg constructor becomes something of an issue.

Which is the greater priority? Keeping classes safe and coherent at the cost of changing the API when something becomes risky, or keeping the API intact at the cost of making classes more fragile than they could be?